hydromt/raster.py

```diff
@@ -520,21 +545,34 @@ def res(self) -> tuple[float, float]:
 
         NOTE: rotated rasters with a negative dx are not supported.
         """
-        xs, ys = self.xcoords.data, self.ycoords.data
-        dx, dy = 0, 0
-        if xs.ndim == 1:
-            dx = xs[1] - xs[0]
-            dy = ys[1] - ys[0]
-        elif xs.ndim == 2:
-            ddx0 = xs[1, 0] - xs[0, 0]
-            ddy0 = ys[1, 0] - ys[0, 0]
-            ddx1 = xs[0, 1] - xs[0, 0]
-            ddy1 = ys[0, 1] - ys[0, 0]
-            dx = math.hypot(ddx1, ddy1)  # always positive!
-            dy = math.hypot(ddx0, ddy0)
-            rot = self.rotation
+        if self._res is not None:  # cached resolution
+            return self._res
+        elif self._transform is not None:  # cached transform
+            if self._transform.b == self._transform.d == 0:
+                dx, dy = self._transform.a, self._transform.e
+            else:  # rotated
+                xy0 = self._transform * (0, 0)
+                x1y = self._transform * (1, 0)
+                xy1 = self._transform * (0, 1)
+                ddx0, ddy0 = xy1[0] - xy0[0], xy1[1] - xy0[1]
+                dx = math.hypot(x1y[0] - xy0[0], x1y[1] - xy0[1])
+                dy = math.hypot(xy1[0] - xy0[0], xy1[1] - xy0[1])
+        else:  # from coordinates
+            xs, ys = self.xcoords.data, self.ycoords.data
+            if xs.ndim == 1:
+                dx = xs[1] - xs[0]
+                dy = ys[1] - ys[0]
+            elif xs.ndim == 2:
+                ddx0 = xs[1, 0] - xs[0, 0]
+                ddy0 = ys[1, 0] - ys[0, 0]
+                ddx1 = xs[0, 1] - xs[0, 0]
+                ddy1 = ys[0, 1] - ys[0, 0]
+                dx = math.hypot(ddx1, ddy1)  # always positive!
+                dy = math.hypot(ddx0, ddy0)
+        rot = self.rotation
+        if rot != 0:
```

Comment:
if `rot` is a float and not an integer this is very error prone. In that case please use `np.isclose`